### PR TITLE
[Chore] Upgrade guide tracker for released and unreleased features

### PR DESCRIPTION
**Pull request by art-alexeyenko**
- [x] This PR follows the [Contribution Guide](https://github.com/Sitecore/jss/blob/dev/CONTRIBUTING.md)
- [ ] Changelog updated

## Description / Motivation
Track future required upgrade steps with this one simple trick (file). 
Adds a file to track unreleased changes to templates and add-ons (`JSS next` section) and upgrade instructions for released versions. The goal is to have a detailed list of changes for an upgrade guide ready before each release. 
Also modifies PR checklist. Each PR should include an entry to UPGRADING.md when applicable.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Chore

### CODE REVIEW

File: docs/UPGRADING.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# Upgrade guides
+
+This document includes the list of required changes when upgrading an existing JSS app.
+Each version contains the list of files, dependencies and environment variables that need to be applied or updated. Unreleased changes are listed in [JSS Next](#jss-next).
```
Comment (kendoce) [inline docs/UPGRADING.md:4]: a bit misleading for a first time reader. Can we just name it as Unreleased just like in our Changelog?

File: docs/UPGRADING.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+Changes are grouped by templates and add-ons.
+
+## JSS Next
+
+### nextjs
```
Comment (kendoce) [inline docs/UPGRADING.md:10]: Do we need these now? if so, can we rename it like the headers in the upgrade doc(i.e. [Update SXA add-on files](https://doc.sitecore.com/xmc/en/developers/jss/216/jss-xmc/upgrade-jss-21-5-next-js-apps-to-version-21-6.html#:~:text=nextjs/utils%27%3B%20statement.-,Update%20SXA%20add%2Don%20files,-To%20upgrade%20the)?

**Review by kendoce — CHANGES_REQUESTED** [review]

**Review by illiakovalenko — CHANGES_REQUESTED** [review]
According to work item, our decision/idea was to create an "Upgrade guide" document per version
`docs/upgrading/<version number>.md`
I would prefer to avoid having a huge blanket of upgrades in a single file
It should be really easy to read and quickly check all the changes (also find previous changes if they are needed for customer), think about yourself as a customer.
Take a look at this [example](https://github.com/Laravel-Backpack/Docs/blob/master/6.x/upgrade-guide.md), you can find other examples in a network
cc @kendoce

**Comment by illiakovalenko** [comment]
@kendoce, @art-alexeyenko 
Another idea, may be it might be better even, need to explore
There is a way to create a guide under _wiki_ tab that is more appropriate than an .md file
See [example](https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/wiki/%5BUser-Guide%5D-Upgrade-from-v3.3.0-to-v4.0.0)

**Comment by art-alexeyenko** [comment]
@illiakovalenko the decision to make a single .md file for this is deliberate, for a few reasons:
1. Maintainability. We can realistically only provide update instructions incrementally (21.7->21.8, 21.8->22.0 etc). If we are doing .md files we can quickly get a bloated folder that is annoying to find correct files in
2. Findability. It would also be annoying for customers to switch between files in a large folder, it's much easier to do a ctrl+F to find a relevant version in a single file

Wiki is worth exploring though, thank you for the idea

File: docs/UPGRADING.md (hunk 2 above)
Comment (art-alexeyenko) [reply docs/UPGRADING.md:10]: for work-in-progress entries we do need the per-tempalte sections, makes it easier to add and sort entries, less chace a change is lost or mislabeled. They can be cleaned up when doing release activities.

**Comment by art-alexeyenko** [comment]
@illiakovalenko please check this updated structure.
It should make the docs more orderly, while not complicating the navigation too much.

I explored the wiki option. It seems wikis are hosted as a separate repo and we'd need a separate PR or push to updates them. I think wikis would be a better option with infrequent updates, but our aim here is to keep track of update guides every other PR or so. Plus, it's easier for community contributors to list the changes with current approach, if they choose to.

For easier maintenance, do you think the gitignore file in release branches should be updated to not track the docs folder? Reduce potential conflicts when cherrypicking changes and keep the guides in one location only?
CC @kendoce

**Review by illiakovalenko — CHANGES_REQUESTED** [review]
Looks better!
I think regarding _.gitignore_ we can exclude only _unreleased.md_ file.
And we need to update our publish scripts to make _unreleased.md_ file empty in _dev_ branch, when we publish a new minor/major canary in _dev_ branch

File: docs/upgrades/21.x/21.7.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+## Upgrade to JSS 21.7 from 21.6
+
+* Find the list of changes in Sitecore documentation:
+  * [XM Cloud (to be added when published)]()
```
[the remark this answers is not in the document]
Comment (art-alexeyenko) [reply docs/upgrades/21.x/21.7.md:4]: Thanks! Could not find it yesterday.

File: UPGRADING.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# Upgrade guides
+
+This document includes the links to upgrade guides for different versions of JSS apps.
+Each version contains the list of files, dependencies and environment variables that need to be added, updated or removed. Changes are grouped by templates and add-ons in each corresponding page.
+Work-in-progress items are listed in [Unreleased](./docs/upgrades/unreleased.md).
+
+
```
[the remark this answers is not in the document]
Comment (illiakovalenko) [reply UPGRADING.md:7]: This line needs to be removed

**Review by illiakovalenko — APPROVED** [review]
Reopened one comment, feel free to push another minor change and merge after that 👍

**Pull request merged**